src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -3346,6 +3358,176 @@ public AttributeProviderPair (CustomAttribute attribute, ICustomAttributeProvide
+						// static CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder? binder, object?[]? args, System.Globalization.CultureInfo? culture)
+						// static CreateInstance (System.Type type, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder? binder, object?[]? args, System.Globalization.CultureInfo? culture, object?[]? activationAttributes) { throw null; }
+						//
+						case "CreateInstance" when !calledMethod.ContainsGenericParameter
+							&& calledMethod.DeclaringType.Name == "Activator"
+							&& calledMethod.Parameters.Count >= 1
+							&& calledMethod.Parameters[0].ParameterType.MetadataType != MetadataType.String: {
+
+								var parameters = calledMethod.Parameters;
+
+								reflectionContext.AnalyzingPattern ();
+
+								int? ctorParameterCount = null;
+								BindingFlags bindingFlags = BindingFlags.Instance;
+								if (parameters.Count > 1) {									
+									if (parameters [1].ParameterType.MetadataType == MetadataType.Boolean) {
+										// The overload that takes a "nonPublic" bool
+										bool nonPublic = true;
+										if (methodParams [1] is ConstIntValue constInt) {
+											nonPublic = constInt.Value != 0;
+										}
+
+										if (nonPublic)
+											bindingFlags |= BindingFlags.NonPublic | BindingFlags.Public;
+										else
+											bindingFlags |= BindingFlags.Public;
+										ctorParameterCount = 0;
+									} else {
+										// Overload that has the parameters as the second or fourth argument
+										int argsParam = parameters.Count == 2 || parameters.Count == 3 ? 1 : 3;
+										
+										if (methodParams.Count > argsParam &&
+											methodParams [argsParam] is ArrayValue arrayValue &&
+											arrayValue.Size.AsConstInt () != null) {
+											ctorParameterCount = arrayValue.Size.AsConstInt ();
+										}
+
+										if (parameters.Count > 3) {
+											if (methodParams [1].AsConstInt () != null)
+												bindingFlags |= (BindingFlags)methodParams [1].AsConstInt ();
+											else
+												bindingFlags |= BindingFlags.NonPublic | BindingFlags.Public;
+										} else {
+											bindingFlags |= BindingFlags.Public;
+										}
+									}
+								}
+								else {
+									// The overload with a single System.Type argument
+									ctorParameterCount = 0;
+									bindingFlags |= BindingFlags.Public;
+								}
+
+								// Go over all types we've seen
+								foreach (var value in methodParams[0].UniqueValues ()) {
+									if (value is SystemTypeValue systemTypeValue) {
+										MarkMethodsFromReflectionCall (ref reflectionContext, systemTypeValue.TypeRepresented, ".ctor", bindingFlags, ctorParameterCount);
+									} else if (value == NullValue.Instance) {
+										// Nothing to report. This is likely just a value on some unreachable branch.
+										reflectionContext.RecordHandledPattern ();
```

Comment:
Why do we need a new method - why not use the existing RecordRecognizedPattern?

Reply:
This handles:

```csharp
Type t = null;
if (condition)
    t = typeof(Foo);
else
    t = typeof(Bar);
CreateInstance(t);
```

We're going to see `null`, `Foo`, `Bar` for this. We could also report nothing for `null`, but then this is going to assert/crash on `CreateInstance(null)` (I'm sure there will be user code with this, however meaningless that pattern is).

`RecordRecognizedPattern` expects that there's going to be some accessed item. In this case there's no accessed item.

Reply:
In that the correct reporting would be "match - Foo" and "match - Bar", ignoring the `null` completely. But if we see only `null` (alone) we should report it is "unrecognized". That's how the rest of the pattern matching behaves in cases where there are potentially multiple matches (like method overrides for `GetMethod` for example).

Reply:
I tried to update the `if` cascade to make it like you suggest (report CreateInstance(null) as unrecognized, even though it's recognized), but things got awkward quickly. I think we'll want to make make adjustments to the reflection context anyway to make it more convenient to use it at merge points (where we have multiple values, some of which might be known).

I added that to the running list of issues to look into.